[Core] fix price calculation with immutables and adjustments

dpfaffenbauer opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  |no
| BC breaks?    | no
| Deprecations? |no
| Fixed tickets | #2572

<!--
Write a short README entry for your feature/bugfix here (replace this comment block.)
This will help people understand your PR and can be used as a start of the Doc PR.
-->

dpfaffenbauer committed:
[Core] fix price calculation with immutables and adjustments

jdreesen commented on src/CoreShop/Component/Core/Order/Processor/CartSubtotalProcessor.php:
```diff
@@ -33,8 +33,21 @@ public function process(OrderInterface $cart): void
         foreach ($cart->getItems() as $item) {
-            $subtotalGross += $item->getTotal(true);
-            $subtotalNet += $item->getTotal(false);
+            /**
```
This is a comment and not a DocBlock, right?
```suggestion
            /*
```

jdreesen commented on src/CoreShop/Component/Core/Order/Processor/CartItemsImmutableProcessor.php:
```diff
@@ -68,6 +68,14 @@ public function process(OrderInterface $cart): void
                 $context
             );
+
+            /**
+             * https://github.com/coreshop/CoreShop/issues/2572
+             *
+             * Since we are recalculating the items-total, we also need to respect the adjustments and
+             * re-add them to the item total as well.
+            */
```
```suggestion
            /*
             * https://github.com/coreshop/CoreShop/issues/2572
             *
             * Since we are recalculating the items-total, we also need to respect the adjustments and
             * re-add them to the item total as well.
             */
```

dpfaffenbauer committed:
Update src/CoreShop/Component/Core/Order/Processor/CartItemsImmutableProcessor.php

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

dpfaffenbauer committed:
Update src/CoreShop/Component/Core/Order/Processor/CartSubtotalProcessor.php

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

solverat commented:
@dpfaffenbauer thanks for the quick response. Unfortunately, this does not work when using additional adjustments on cart items. In other words: The total value of the order items are getting completely ignored now.

To reproduce: You may want to add a simple adjustment to an item (not cart) in your test to validate the total value. 

I'm not sure why this PR is not failing, but I guess it's because `CartRuleApplier` is adding an adjustment on item- AND cart layer:

https://github.com/coreshop/CoreShop/blob/1ca9f58651ebb314a56a942d5c7e1df5e89eb423/src/CoreShop/Component/Core/Cart/Rule/Applier/CartRuleApplier.php#L189-L190

https://github.com/coreshop/CoreShop/blob/1ca9f58651ebb314a56a942d5c7e1df5e89eb423/src/CoreShop/Component/Core/Cart/Rule/Applier/CartRuleApplier.php#L200-L202

dpfaffenbauer commented:
@solverat adding an adjustment to the item without the order is not supported and is something we cannot make work easily without breaking other things. as you pointed out, cart-rules are applied on cart-item and cart directly. and that is the reason it is applied on both.

dpfaffenbauer commented:
@solverat one solution could be that we set the adjustment on the cart itself as neutral and use them from the items. not sure what side-effects that can have.

dpfaffenbauer committed:
[Core] make cart-price-rule items adjustment neutral and add a calculateSubtotal function

dpfaffenbauer commented:
@solverat please check the latest changes

dpfaffenbauer committed:
[Core] make cart-price-rule-items adjustment on order neutral

solverat commented:
@dpfaffenbauer This looks VERY promising! 

In my Test, I've
- added a 2% global cart discount rule
- added a global surcharge cart rule with fix amount
- added a custom adjustment (just as before, only on item layer but with an `isImmutable` check)

…and all calculations are correct now!

The only thing I don't understand in your PR:

https://github.com/coreshop/CoreShop/pull/2573/commits/9e742c8660ef099ca0662eae6c5f95a6d8198cbd#diff-ad51a1ab22ce6a03fe767ee679b9ccfd41a00b017606df746e1baf50b4c5f490R61-R72

I removed that part, and I wasn't able to spot the difference. Maybe a hint to test it would be great!

dpfaffenbauer committed:
[Core] make cart-price-rule-items adjustment on order neutral

dpfaffenbauer commented:
@solverat If you have already have items with a total of 0, without that, it would lead to a division by 0 error.

I did some other changes as well, can you test them as well? Since we are now changing which discounts apply to what, we need to make sure to always have the right calculations.

dpfaffenbauer committed:
[Core] make cart-price-rule-items adjustment on order neutral

solverat commented:
@dpfaffenbauer Just adopted your changes: Still valid!

dpfaffenbauer commented:
I will also test it in one of our projects first. It might have unconsidered side-effects

solverat commented:
I also think that this PR makes the neutral adjustment calculation more solid. :+1: 

It's quite clear now, what's happening when adding custom adjustments - especially if they have to be added to the order item subtotal to allow them to be discountable by cart rules!

dpfaffenbauer commented:
Yeah, so basically what this changes is:
 - Cart Price Rules are applied to the Cart Total, and added to the Cart Item as Neutral Adjustment
 - Cart Price Item Rules are applied to the Cart Item Total, and added to the Cart as Neutral Adjustment

solverat commented:
@dpfaffenbauer Btw, I was not able to test cart items action (`coreshop.cart_item_price_rule.action`). They never get applied (When using discount amount for example).

I don't know where the cart item rule processor should get triggered?

dpfaffenbauer commented:
@solverat They do work though for me. They are triggered as Cart Item Rule Action.

solverat commented:
On my debug marker, only `CoreShop\Component\Core\Cart\Rule\Action\DiscountAmountActionProcessor` gets dispatched, but never `CoreShop\Component\Core\CartItem\Rule\Action\DiscountAmountActionProcessor`.

dpfaffenbauer commented:
@solverat 
![Screenshot 2024-03-05 at 10 40 26](https://github.com/coreshop/CoreShop/assets/5981845/31b7dc4d-9fe0-4f3c-9acd-2b44a7284bcc)
![Screenshot 2024-03-05 at 10 40 31](https://github.com/coreshop/CoreShop/assets/5981845/d0855121-b040-47d2-8567-fd4fd053570a)
![Screenshot 2024-03-05 at 10 42 11](https://github.com/coreshop/CoreShop/assets/5981845/4cbbfda5-0325-4baf-bb70-b5e985661145)

solverat commented:
:see_no_evil: I'm so sorry, I'm very new to coreshop...
